File: apps/mobile/lib/features/habits/domain/habit.dart
class HabitLog {
  const HabitLog({
    required this.id,
    required this.habitId,
    required this.loggedOn,
    required this.count,
    required this.isCompleted,
    this.value,
  });

  final String id;
  final String habitId;
  final DateTime loggedOn;
  final int count;
  final bool isCompleted;
  final double? value;

  factory HabitLog.fromJson(Map<String, dynamic> json) => HabitLog(
        id: json['id'] as String,
        habitId: json['habit_id'] as String,
        loggedOn: DateTime.parse(json['logged_on'] as String),
        count: json['count'] as int? ?? 0,
        isCompleted: json['is_completed'] as bool? ?? false,
        value: (json['value'] as num?)?.toDouble(),
      );

  Map<String, dynamic> toJson() => <String, dynamic>{
        'id': id,
        'habit_id': habitId,
        'logged_on': loggedOn.toIso8601String().split('T').first,
        'count': count,
        'is_completed': isCompleted,
        'value': value,
      };
}

class Habit {
  const Habit({
    required this.id,
    required this.name,
    required this.icon,
    required this.frequency,
    required this.targetCount,
    required this.currentStreak,
    required this.longestStreak,
    this.color,
    this.targetValue,
    this.unit,
    this.activeWeekdays = const <int>[],
    this.reminderTime,
    this.isArchived = false,
    this.position = 0,
    this.today,
    this.completionRate30d = 0,
  });

  final String id;
  final String name;
  final String icon;
  final String frequency;
  final int targetCount;
  final int currentStreak;
  final int longestStreak;
  final String? color;
  final double? targetValue;
  final String? unit;
  final List<int> activeWeekdays;
  final String? reminderTime;
  final bool isArchived;
  final int position;
  final HabitLog? today;
  final double completionRate30d;

  bool get isDoneToday => today?.isCompleted ?? false;

  int get todayCount => today?.count ?? 0;

  factory Habit.fromJson(Map<String, dynamic> json) => Habit(
        id: json['id'] as String,
        name: json['name'] as String,
        icon: json['icon'] as String? ?? 'check_circle',
        frequency: json['frequency'] as String? ?? 'daily',
        targetCount: json['target_count'] as int? ?? 1,
        currentStreak: json['current_streak'] as int? ?? 0,
        longestStreak: json['longest_streak'] as int? ?? 0,
        color: json['color'] as String?,
        targetValue: (json['target_value'] as num?)?.toDouble(),
        unit: json['unit'] as String?,
        activeWeekdays: ((json['active_weekdays'] as List<dynamic>?) ?? const <dynamic>[])
            .map((dynamic item) => item as int)
            .toList(),
        reminderTime: json['reminder_time'] as String?,
        isArchived: json['is_archived'] as bool? ?? false,
        position: json['position'] as int? ?? 0,
        today: json['today'] == null
            ? null
            : HabitLog.fromJson(
                Map<String, dynamic>.from(json['today'] as Map<dynamic, dynamic>),
              ),
        completionRate30d: (json['completion_rate_30d'] as num?)?.toDouble() ?? 0,
      );

  Map<String, dynamic> toJson() => <String, dynamic>{
        'id': id,
        'name': name,
        'icon': icon,
        'frequency': frequency,
        'target_count': targetCount,
        'current_streak': currentStreak,
        'longest_streak': longestStreak,
        'color': color,
        'target_value': targetValue,
        'unit': unit,
        'active_weekdays': activeWeekdays,
        'reminder_time': reminderTime,
        'is_archived': isArchived,
        'position': position,
        'today': today?.toJson(),
        'completion_rate_30d': completionRate30d,
      };
}

File: services/api/app/schemas/habit.py
"""Habit schemas."""

from __future__ import annotations

from datetime import date, datetime

from pydantic import Field

from app.models.enums import HabitFrequency
from app.schemas.common import APIModel


class HabitCreate(APIModel):
    name: str = Field(..., min_length=1, max_length=120)
    icon: str = Field("check_circle", max_length=40)
    color: str | None = Field(None, pattern="^#[0-9A-Fa-f]{6}$")
    frequency: HabitFrequency = HabitFrequency.DAILY
    target_count: int = Field(1, ge=1, le=50)
    target_value: float | None = Field(None, gt=0, le=100000)
    unit: str | None = Field(None, max_length=24)
    active_weekdays: list[int] = Field(default_factory=list)
    reminder_time: str | None = Field(None, pattern="^([01]\\d|2[0-3]):[0-5]\\d$")


class HabitUpdate(APIModel):
    name: str | None = Field(None, min_length=1, max_length=120)
    icon: str | None = Field(None, max_length=40)
    color: str | None = Field(None, pattern="^#[0-9A-Fa-f]{6}$")
    frequency: HabitFrequency | None = None
    target_count: int | None = Field(None, ge=1, le=50)
    target_value: float | None = Field(None, gt=0, le=100000)
    unit: str | None = Field(None, max_length=24)
    active_weekdays: list[int] | None = None
    reminder_time: str | None = Field(None, pattern="^([01]\\d|2[0-3]):[0-5]\\d$")
    is_archived: bool | None = None
    position: int | None = Field(None, ge=0)


class HabitLogWrite(APIModel):
    logged_on: date
    count: int = Field(1, ge=0, le=50)
    value: float | None = Field(None, ge=0, le=100000)
    client_uuid: str | None = Field(None, max_length=64)


class HabitLogRead(APIModel):
    id: str
    habit_id: str
    logged_on: date
    count: int
    value: float | None
    is_completed: bool


class HabitRead(APIModel):
    id: str
    name: str
    icon: str
    color: str | None
    frequency: HabitFrequency
    target_count: int
    target_value: float | None
    unit: str | None
    active_weekdays: list[int] = Field(default_factory=list)
    reminder_time: str | None
    is_archived: bool
    position: int
    current_streak: int
    longest_streak: int
    created_at: datetime
    #: Today's log, when one exists — lets the list render without an N+1.
    today: HabitLogRead | None = None
    completion_rate_30d: float = 0.0
